### PR TITLE
Revert backwards incompatible changes

### PR DESCRIPTION
**Pull request by apelisse**
Can we live with that? Everything still passes, but at least we don't have to bump the major, so any backport/cherry-pick will be much easier to sell.

### CODE REVIEW

File: schema/elements.go [hunk 1]
```diff
@@ -30,7 +30,11 @@ type Schema struct {
 	once sync.Once
 	m    map[string]TypeDef
 
-	lock sync.Mutex
+	// Once used to protect the initialization of `lock` field.
+	lockOnce sync.Once
+	// Lock which protects writes to resolvedTypes. Used as pointer so that
+	// schema may be used as a value type
+	lock *sync.Mutex
```
Comment (apelisse) [inline schema/elements.go:37]: Do we need all these locks here? Can we make the lock by value since we're not supposed to copy a schema?

Comment (alexzielenski) [reply schema/elements.go:37]: yeah there were no unsafe schema usages in the ptr semantics refactor

Comment (alexzielenski) [reply schema/elements.go:37]: can remove the once. just need the lock

**Comment by alexzielenski** [comment]
/hold

tentative lgtm

**Comment by apelisse** [comment]
OK I've tried to fix the tags, done this:
- Created [`v4.2.3`](https://github.com/kubernetes-sigs/structured-merge-diff/tree/v4.2.3) with this merge commit
- Removed `release-5.0` branch and created [`release-5.0-wrong`](https://github.com/kubernetes-sigs/structured-merge-diff/tree/v5.0.0-wrong) instead
- Created [`v4.2.2-wrong`](https://github.com/kubernetes-sigs/structured-merge-diff/tree/v4.2.2-wrong) to clarify why we skipped that version

**Pull request merged**